File: BTinorderpostorder.py
"""
The apporach here is same. first find the root element from the postorder. it would be the last element
in the postorder array. now find for that element in the inorder, the left elements towards the root element
found would be the elements towards the left of the main-tree and the elements towards the right will
be the right nodes of the tree. now create new inorder sub-arrays for each left and right sub trees by passing
appropriate indices mentioned. call the function recursively to build the entire tree.
Leetcode - running
Time complexity - O(N)
Space complexity - O(N)
"""
def buildTree(self, inorder, postorder):
        if len(inorder) ==0 or len(postorder) ==0:
            return None
        val = postorder.pop(-1)
        root = TreeNode(val)
        mid = inorder.index(val)
        root.right = self.buildTree(inorder[mid+1:len(inorder)],postorder)
        root.left = self.buildTree(inorder[0:mid],postorder)
        return root

        

File: bt_inorder_postorder.java
// Time Complexity : O(n) 
// Space Complexity : O(n) 
// Did this code successfully run on Leetcode : Yes

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    HashMap<Integer, Integer> map = new HashMap<>();
    int index;  //post order index
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        if(inorder == null)
            return null;
        for(int i=0; i<inorder.length; i++)
            map.put(inorder[i], i);
        index = postorder.length - 1;
        return helper(postorder, 0, inorder.length -1);
    }
    private TreeNode helper(int[] postorder, int start, int end){
        if(start > end)
            return null;
        
        int rootValue = postorder[index]; index--;
        TreeNode root = new TreeNode(rootValue);
        int rootIndex = map.get(rootValue);
        root.right = helper(postorder, rootIndex+1, end);
        root.left = helper(postorder, start, rootIndex-1);
        return root;
            
    }
}